File: sdks/pyrax/challenges/upload_folder.py
#!/usr/bin/env python

import os
import pyrax
import time

container_name = 'my-site'
folder_to_upload = os.getenv('TEST_DIRECTORY')
username = os.getenv('RAX_USERNAME')
api_key = os.getenv('RAX_API_KEY')
auth_endpoint = os.getenv('RAX_AUTH_URL')

pyrax.set_setting("identity_type", "rackspace")
# Create the identity object
pyrax._create_identity()
# Change its endpoint
pyrax.identity.auth_endpoint = auth_endpoint + '/v2.0/'

# Identity Connection - Authenticate
pyrax.set_credentials(username, api_key)

cf = pyrax.cloudfiles
print 'Uploading folder'
upload_key, total_bytes = cf.upload_folder(folder_to_upload, container=container_name)
print "Total bytes to upload:", total_bytes
uploaded = 0
while uploaded < total_bytes:
    uploaded = cf.get_uploaded(upload_key)
    print "Progress: %4.2f%%" % ((uploaded * 100.0) / total_bytes)
    time.sleep(1)

print 'Done uploading'

File: sdks/fog/challenges/upload_folder.rb
#!/usr/bin/env ruby

require 'fog'
connection_opts = {
    :rackspace_username   => ENV['RAX_USERNAME'],
    :rackspace_api_key    => ENV['RAX_API_KEY'],
    :rackspace_region     => ENV['RAX_REGION'].downcase.to_sym,
    :rackspace_auth_url   => "#{ENV['RAX_AUTH_URL']}/v2.0"
}
storage = Fog::Storage.new(connection_opts.merge({:provider => 'rackspace'}))

container_name = 'my-site'
folder_to_upload = ENV['TEST_DIRECTORY']
username = ENV['RAX_USERNAME']
api_key = ENV['RAX_API_KEY']
auth_endpoint = ENV['RAX_AUTH_URL']

puts 'Uploading folder'
local_directory = Pathname.new folder_to_upload
remote_directory = storage.directories.create :key => container_name
Dir["#{folder_to_upload}/**.*"].each do |file|
  file_key = Pathname.new(file).relative_path_from local_directory
  remote_directory.files.create :key => file_key.to_s, :body => File.open(file)
end
puts 'Done uploading'
